File: android/app/src/main/java/info/bagen/dwebbrowser/microService/sys/nativeui/NativeUiController.kt
package info.bagen.dwebbrowser.microService.sys.nativeui


import androidx.activity.ComponentActivity
import androidx.compose.runtime.Composable
import androidx.compose.runtime.SideEffect
import androidx.compose.runtime.mutableStateOf
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
import info.bagen.dwebbrowser.microService.sys.nativeui.helper.QueryHelper
import info.bagen.dwebbrowser.microService.sys.nativeui.navigationBar.NavigationBarController
import info.bagen.dwebbrowser.microService.sys.nativeui.safeArea.SafeAreaController
import info.bagen.dwebbrowser.microService.sys.nativeui.statusBar.StatusBarController
import info.bagen.dwebbrowser.microService.sys.nativeui.virtualKeyboard.VirtualKeyboardController

class NativeUiController(
    val activity: ComponentActivity,
) {
    val windowInsetsController by lazy {
        WindowCompat.getInsetsController(
            activity.window, activity.window.decorView
        )
    }
    val currentInsets =
        mutableStateOf(WindowInsetsCompat.toWindowInsetsCompat(activity.window.decorView.rootWindowInsets))

    fun getCurrentInsets(typeMask: Int) = currentInsets.value.getInsets(typeMask)

    val statusBar = StatusBarController(activity, this)
    val navigationBar = NavigationBarController(activity, this)
    val virtualKeyboard = VirtualKeyboardController(activity, this)

    val safeArea = SafeAreaController(activity, this)


    @Composable
    fun effect(): NativeUiController {
        /**
         * 这个 NativeUI 的逻辑是工作在全屏幕下，所以会使得默认覆盖 系统 UI
         */
        SideEffect {
            WindowCompat.setDecorFitsSystemWindows(activity.window, false)
            /// system-bar 一旦隐藏（visible = false），那么被手势划出来后，过一会儿自动回去
            windowInsetsController.systemBarsBehavior = BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE

            ViewCompat.setOnApplyWindowInsetsListener(activity.window.decorView) { _, insets ->
                currentInsets.value = insets
                insets
            }

        }
        statusBar.effect()
        navigationBar.effect()
        virtualKeyboard.effect()
        safeArea.effect()

        return this
    }


    companion object {
        init {
            QueryHelper.init() // 初始化
        }
    }

}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/NativeUI/NativeUiController.cs
using DwebBrowser.MicroService.Browser.Mwebview;
using DwebBrowser.MicroService.Browser.NativeUI.NavigationBar;
using DwebBrowser.MicroService.Browser.NativeUI.SafeArea;
using DwebBrowser.MicroService.Browser.NativeUI.StatusBar;
using DwebBrowser.MicroService.Browser.NativeUI.VirtualKeyboard;

namespace DwebBrowser.MicroService.Browser.NativeUI;

public class NativeUiController
{
    // TODO: 多个Dwebview共用一个MWebviewController，共享nativeui状态？
    // Android是多个Dwebview公用一个Activity，共享nativeui状态？
    public static NativeUiController FromMultiWebView(Mmid mmid) =>
        throw new Exception(string.Format("native ui is unavailable for {0}", mmid));

    public StatusBarController StatusBar { get; init; }
    public NavigationBarController NavigationBar { get; init; }
    public VirtualKeyboardController VirtualKeyboard { get; init; }
    public SafeAreaController SafeArea { get; init; }

    static NativeUiController()
    {
        _ = new QueryHelper();
    }

    public NativeUiController(MultiWebViewController mwebviewController)
    {
        StatusBar = new(mwebviewController, this);
        NavigationBar = new(mwebviewController, this);
        VirtualKeyboard = new(mwebviewController, this);
        SafeArea = new(mwebviewController, this);
    }
}

